src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java

```diff
@@ -0,0 +1,337 @@
+        }
+        for (ShuntCompensator shunt : shuntsToLose) {
+            double nominalV = shunt.getTerminal().getVoltageLevel().getNominalV();
+            shuntIdsToShift.put(shunt.getId(), shunt.getB() * nominalV * nominalV / PerUnit.SB);
+        }
+
+        return new PropagatedContingency(contingency, index, branchIdsToOpen, hvdcIdsToOpen, switchesToOpen,
+                                         generatorIdsToLose, loadIdsToShift, shuntIdsToShift);
+    }
+
+    private static boolean isCoupler(Switch s) {
+        VoltageLevel.NodeBreakerView nbv = s.getVoltageLevel().getNodeBreakerView();
+        Terminal terminal1 = nbv.getTerminal1(s.getId());
+        Terminal terminal2 = nbv.getTerminal2(s.getId());
+        if (terminal1 == null || terminal2 == null) {
+            return false; // FIXME: this can be a coupler, a traverser could be used to detect it
+        }
+        Connectable<?> c1 = terminal1.getConnectable();
+        Connectable<?> c2 = terminal2.getConnectable();
+        return c1 != c2 && c1.getType() == IdentifiableType.BUSBAR_SECTION && c2.getType() == IdentifiableType.BUSBAR_SECTION;
+    }
+
+    public Optional<LfContingency> toLfContingency(LfNetwork network, GraphDecrementalConnectivity<LfBus> connectivity,
+                                                   boolean useSmallComponents) {
+        // find contingency branches that are part of this network
+        Set<LfBranch> branches = new HashSet<>(1);
+        for (String branchId : branchIdsToOpen) {
+            LfBranch branch = network.getBranchById(branchId);
+            if (branch != null) {
+                branches.add(branch);
+            }
+        }
+
+        // check if contingency split this network into multiple components
+        if (branches.isEmpty()
+                && hvdcIdsToOpen.isEmpty()
+                && shuntIdsToShift.isEmpty()
+                && loadIdsToShift.isEmpty()
+                && generatorIdsToLose.isEmpty()) {
+            return Optional.empty();
+        }
+
+        // update connectivity with triggered branches
+        for (LfBranch branch : branches) {
+            connectivity.cut(branch.getBus1(), branch.getBus2());
+        }
+
+        // add to contingency description buses and branches that won't be part of the main connected
+        // component in post contingency state
+        Set<LfBus> buses;
+        if (useSmallComponents) {
+            buses = connectivity.getSmallComponents().stream().flatMap(Set::stream).collect(Collectors.toSet());
+        } else {
+            int slackBusComponent = connectivity.getComponentNumber(network.getSlackBus());
+            buses = network.getBuses().stream().filter(b -> connectivity.getComponentNumber(b) != slackBusComponent).collect(Collectors.toSet());
+        }
+        buses.forEach(b -> branches.addAll(b.getBranches()));
+
+        for (String hvdcId : hvdcIdsToOpen) {
```

Comment:
I added an exception to warm that HVDC line contingency is not yet implemented in SA and sense analysis